modules/gmc-agent/src/main/java/org/gridgain/action/controller/QueryActionsController.java

```diff
@@ -0,0 +1,188 @@
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.MANAGEMENT_POOL;
+import static org.gridgain.utils.QueryUtils.fetchSqlQueryRows;
+import static org.gridgain.utils.QueryUtils.getColumns;
+import static org.gridgain.utils.QueryUtils.prepareQuery;
+
+/**
+ * Query actions controller.
+ */
+@ActionController("QueryActions")
+public class QueryActionsController {
+    /** Context. */
+    private final GridKernalContext ctx;
+
+    /** Query registry. */
+    private final QueryHolderRegistry qryRegistry;
+
+    /** Query process. */
+    private GridQueryProcessor qryProc;
+
+    /** Logger. */
+    private IgniteLogger log;
+
+    /**
+     * @param ctx Context.
+     */
+    public QueryActionsController(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(QueryActionsController.class);
+
+        qryProc = ctx.query();
+        qryRegistry = new QueryHolderRegistry(ctx, Duration.ofMinutes(5));
+    }
+
+    /**
+     * Cancel query by query ID.
+     *
+     * @param qryId Query id.
+     */
+    public CompletableFuture<Void> cancel(String qryId) {
+        qryRegistry.cancelQuery(qryId);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * @param arg Argument.
+     * @return Next page with result.
+     */
+    public CompletableFuture<QueryResult> nextPage(NextPageQueryArgument arg) {
+        final CompletableFuture<QueryResult> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            try {
+                String qryId = arg.getQueryId();
+                if (F.isEmpty(qryId))
+                    throw new IllegalArgumentException("Fail to execute query - query id can't be empty.");
+
+                String cursorId = arg.getCursorId();
+                if (F.isEmpty(cursorId))
+                    throw new IllegalArgumentException("Fail to execute query - cursor id can't be empty.");
+
+                CursorHolder cursorHolder = qryRegistry.findCursor(qryId, cursorId);
+                QueryResult res = fetchResult(qryId, cursorHolder, arg.getPageSize());
+
+                if (!cursorHolder.hasNext())
+                    qryRegistry.closeQueryCursor(qryId, cursorId);
+
+                fut.complete(res);
+            }
+            catch (Throwable e) {
+                fut.completeExceptionally(e);
+            }
+        }, MANAGEMENT_POOL);
+
+        return fut;
+    }
+
+    /**
+     * @param arg Argument.
+     * @return List of query results.
+     */
+    public CompletableFuture<List<QueryResult>> executeSqlQuery(QueryArgument arg) {
+        final CompletableFuture<List<QueryResult>> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            String qryId = arg.getQueryId();
+            qryRegistry.cancelQuery(qryId);
```

Comment:
Why we need to cancel query here?

Reply:
We should close all cursors and cancel already running query if it exists and run new one.

Reply:
We should block execution of query with same queryId

Reply:
Previous query should closed when user try to execute new from the same query tab